### ISSUE TITLE
r2204 - Outlook 2003 causes processor to spike at 100%

### ISSUE THREAD

**Issue by nvdaTrac** [nvaccess/test_nvdaTracImport#214]
**Reported by pmorris on 9 Jul 2008 02:00 UTC**
OS: Windows XP Service Pack 3
NVDA: trunk-r2204

Observed:
When the user opens a mail message in Outlook 2003 the processor peaks at 100% and NVDA has to be killed to recover system functionality.

The NVDA log does not appear to display any errors when this occurs.

Steps to Repro:
1.  Install  trunk-r2204
2.  After installation, open Microsoft Outlook 2003
3.  Select a mail message.
4.  The mail message opens and NVDA announces the new Window taking focus.
4.  Nothing then appears to happen and NVDA stops functioning.
5.  If the user then opens the Windows Task Manager, the processor is spiked at 100%

**Comment by nvdaTrac** [comment]
**Comment by pmorris on 9 Jul 2008 02:16 UTC**
Additional:  This appears to be a memory bleed.  I monitored the Task Manager and it was over 100M of memory and climbing

**Comment by nvdaTrac** [comment]
**Comment by pmorris on 9 Jul 2008 02:49 UTC**
I closed out of Outlook and the following errors showed in the log:

ERROR - stderr (20:42:21):
Exception
ERROR - stderr (20:42:21):
exceptions
ERROR - stderr (20:42:21):
.
ERROR - stderr (20:42:21):
WindowsError
ERROR - stderr (20:42:21):
:
ERROR - stderr (20:42:21):
'exception: access violation reading 0x00000000'
ERROR - stderr (20:42:21):
 in
ERROR - stderr (20:42:21):
<bound method POINTER(IAccessible).__del__ of <POINTER(IAccessible) ptr=0x199c54 at 20bec10>>
ERROR - stderr (20:42:21):
 ignored

**Comment by nvdaTrac** [comment]
**Comment by jteh on 19 Jul 2008 04:46 UTC**
How big was the message you were trying to load? It is possible that this was not a crash, but rather, just the old virtual buffer code taking a very long time to load. This would explain the CPU spike. How long have you waited before killing NVDA?

Try loading a very small message.

**Comment by nvdaTrac** [comment]
**Comment by pmorris on 20 Jul 2008 16:34 UTC**
I tried this again with the latest snapshot (r2275).

I opened a plain text message that contained the text:
This is a test

This is a test.

NVDA failed to respond and read the message after 19 minutes - at which point I killed the process.

At 16 minutes elapsed NVDA was using 106mb of memory.
At 19 minutes elapsed NVDA was using 131MB of memory.

**Comment by nvdaTrac** [comment]
**Comment by vtsaran on 30 Jul 2008 21:10 UTC**
The problem with Outlook 2003 and text/HTML messages still persists. In looking at the log, it seems that MSHTML.py produces a call for an unknown attribute isContentEditable. After it fails, it seems like some events get stuck in the queue.
Here is couple of lines from the log:

INFO - appModuleHandler.update (14:02:13):
Loaded appModule outlook
ERROR - queueHandler.flushQueue (14:02:34):
Error in func _queueEventCallback from eventQueue
Traceback (most recent call last):
  File "C:\Documents and Settings\vtsaran\My Documents\sources\nvda\source\queueHandler.py", line 53, in flushQueue
    func(*args,**kwargs)
  File "C:\Documents and Settings\vtsaran\My Documents\sources\nvda\source\eventHandler.py", line 40, in _queueEventCallback
    executeEvent(eventName,obj)
  File "C:\Documents and Settings\vtsaran\My Documents\sources\nvda\source\eventHandler.py", line 73, in executeEvent
    executeEvent_appModuleLevel(eventName,obj)
  File "C:\Documents and Settings\vtsaran\My Documents\sources\nvda\source\eventHandler.py", line 111, in executeEvent_appModuleLevel
    executeEvent_defaultAppModuleLevel(name,obj)
  File "C:\Documents and Settings\vtsaran\My Documents\sources\nvda\source\eventHandler.py", line 118, in executeEvent_defaultAppModuleLevel
    executeEvent_virtualBufferLevel(name,obj)
  File "C:\Documents and Settings\vtsaran\My Documents\sources\nvda\source\eventHandler.py", line 125, in executeEvent_virtualBufferLevel
    executeEvent_NVDAObjectLevel(name,obj)
  File "C:\Documents and Settings\vtsaran\My Documents\sources\nvda\source\eventHandler.py", line 129, in executeEvent_NVDAObjectLevel
    getattr(obj,'event_%s'%name)()
  File "C:\Documents and Settings\vtsaran\My Documents\sources\nvda\source\NVDAObjects\IAccessible\MSHTML.py", line 174, in event_gainFocus
    if self.isContentEditable:
  File "C:\Documents and Settings\vtsaran\My Documents\sources\nvda\source\baseObject.py", line 23, in __getattr__
    return getattr(self,'_get_%s'%name)()
  File "C:\Documents and Settings\vtsaran\My Documents\sources\nvda\source\NVDAObjects\IAccessible\MSHTML.py", line 168, in _get_isContentEditable
    if hasattr(self,'domElement') and self.domElement.isContentEditable:
  File "C:\Python25\lib\site-packages\win32com\client\dynamic.py", line 496, in __getattr__
    raise AttributeError, "%s.%s" % (self._username_, attr)
AttributeError: <unknown>.isContentEditable

**Comment by nvdaTrac** [comment]
**Comment by mdcurran on 30 Jul 2008 22:57 UTC**
Jamie and I are currently writing and testing a module for NVDA to take control of COM calls (to MSAA etc) that seem to block for ever. Using this module with NVDA and MS Outlook does stop a lot of the freezes, including this one. However, it does majorly stuff up the old virtualBuffers, so using this module with HTML messages isn't that good. However, once the in-process support for Internet Explorer_Server windows is written in with the new virtualBuffer library, then this will no longer be an issue.
So in short we are working on the issue, and things look promising. But this module really can't be used to fix these problems until MSHTML is supported in the new virtualBuffers.

**Comment by nvdaTrac** [comment]
**Comment by vtsaran on 6 Oct 2008 05:43 UTC**
While the new module is being worked on, is it possible to revert to the old behavior so that we can at least use Outlook 2003? At the moment none of the messages will read with NVDA.

**Comment by nvdaTrac** [comment]
**Comment by vtsaran on 21 Jan 2009 19:33 UTC**
Are there any updates on this sissue? Thanks.

**Comment by nvdaTrac** [comment]
**Comment by psuptic on 6 Apr 2009 20:55 UTC**
Using 2806 here, I can add a bit more information.  If the message being viewed is Rich Text, NVDA reads it perfectly.  As soon as a Plain Text or HTML message is opened, NVDA stops responding and goes to about 50% processor usage.  If I close that message and open any other Rich Text message, NVDA processor use will drop back to normal and it will begin to read.

**Comment by nvdaTrac** [comment]
**Comment by vtsaran on 25 May 2009 17:32 UTC**
Confirmed here as well with the latest snapshot.

**Comment by nvdaTrac** [comment]
**Comment by jteh on 7 Aug 2009 05:36 UTC**
Please try with a recent snapshot. We recently merged some code which may well fix (or at least improve) this situation.

**Comment by nvdaTrac** [comment]
**Comment by jteh on 4 Sep 2009 01:56 UTC**
No comments from affected users, so we can't confirm whether this is fixed.

**Comment by nvdaTrac** [comment]
**Comment by Keffi on 24 Oct 2009 09:41 UTC**
this problem still exists. Reading of text and html messages hangs up NVDA and boosts the CPU. Using NVDA 2009 beta 1

**Comment by nvdaTrac** [comment]
**Comment by mdcurran on 7 Dec 2009 03:02 UTC**
Problem still exists with 2009.1.
On the messages that cause this, its not just opening the message that makes NVDA freeze, Even starting NVDA and then moving focus to the already opened message causes NVDA to freeze. 
Plain text mesages that cause this freeze use the window class RichEdit20W. So its clear it has nothing to do with virtualBuffers, but because it also happens for HTML messages with a virtualBuffer, its not just specific to RichEdit20W either.
When this freeze occures, its still possible to use Outlook, its only NVDA that freezes. This suggests that its not due to part of Outlook freezing because of the Message Window appearing, but instead I think it may be something very broken with the MSAA implementation of one of the message control's ancestor windows.
Though also due to a previous comment about memory leaks, its possible  there is a loop either in our code, or in the MSAA implementation causing our code to keep going around and around.

**Comment by nvdaTrac** [comment]
**Comment by jteh on 7 Dec 2009 03:06 UTC**
Might be worth running traceme on this. That'll confirm whether it is a loop quickly enough and/or might give some indication of where the problem lies.

**Comment by nvdaTrac** [comment]
**Comment by mdcurran on 8 Dec 2009 07:28 UTC**
Technically fixed in r3427. Now rather than freezing, NVDA  pauses for about half a second and then plays the error sound, and from then on works normally.
We can probably fix this even further but for now  some safety code has been added to stop NVDA from going in to an infinit loop when moving up a bad implementation of the focus ancestry in the underlying accessibility hierarchy.

Won't close the ticket though until the underlying problem is solved.

The reason for the freeze:
Outlook displays plain text messages in a RichEdit20W control. Its client (edit) MSAA control's parent is the standard Window MSAA control. However the Window MSAA control's parent seems to have been changed inappropriately, so that it is the client (edit) MSAA control, rather than the client of the parent window. This means that there is a loop in the hierarchy, if you keep going accParent from the RichEdit20W control, you will never get out of it.
When Outlook displays HTML messages, it uses an Internet Explorer_server control. The problem is much the same in that the accParent of the client MSAA control is changed so that it points to itself, again making a loop.

**Comment by nvdaTrac** [comment]
**Comment by mdcurran on 4 Jan 2010 04:53 UTC**
Fixed in r3451. Now the 0.5 second delay and error sound are also gone.
People please test this a lot with Outlook 2003. If it still seems to happen in these situations I.E. the code isn't matching on some systems, please re-open.
